src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Renamer.cs

```diff
@@ -75,55 +77,81 @@ internal class Renamer
 
         public async Task RenameAsync(Project myNewProject)
         {
-            Document newDocument = (from d in myNewProject.Documents where StringComparers.Paths.Equals(d.FilePath, _newFilePath) select d).FirstOrDefault();
-            if (newDocument == null)
+            Solution renamedSolution = await GetRenamedSolutionAsync(myNewProject).ConfigureAwait(false);
+            if (renamedSolution == null)
                 return;
 
-            var root = await newDocument.GetSyntaxRootAsync().ConfigureAwait(false);
-            if (root == null)
-                return;
+            await _threadingService.SwitchToUIThread();
+            var renamedSolutionApplied = _roslynServices.ApplyChangesToSolution(myNewProject.Solution.Workspace, renamedSolution);
 
-            string oldName = Path.GetFileNameWithoutExtension(_oldFilePath);
-            string newName = Path.GetFileNameWithoutExtension(newDocument.FilePath);
+            if (!renamedSolutionApplied)
+            {
+                string failureMessage = string.Format(Resources.RenameSymbolFailed, Path.GetFileNameWithoutExtension(_oldFilePath));
+                await _threadingService.SwitchToUIThread();
+                _userNotificationServices.NotifyFailure(failureMessage);
+            }
+        }
 
-            var declaration = root.DescendantNodes().Where(n => HasMatchingSyntaxNode(newDocument, n, oldName)).FirstOrDefault();
-            if (declaration == null)
-                return;
 
-            var semanticModel = await newDocument.GetSemanticModelAsync().ConfigureAwait(false);
-            if (semanticModel == null)
-                return;
+        private async Task<Solution> GetRenamedSolutionAsync(Project myNewProject)
+        {
+            var project = myNewProject;
+            Solution renamedSolution = null;
+            string oldName = Path.GetFileNameWithoutExtension(_oldFilePath);
 
-            var symbol = semanticModel.GetDeclaredSymbol(declaration);
-            if (symbol == null)
-                return;
+            while (true)
+            {
+                Document newDocument = (from d in project.Documents where StringComparers.Paths.Equals(d.FilePath, _newFilePath) select d).FirstOrDefault();
+                if (newDocument == null)
+                    return renamedSolution;
+                
+                var root = await newDocument.GetSyntaxRootAsync().ConfigureAwait(false);
+                if (root == null)
+                    return renamedSolution;
+
+                var declarations = root.DescendantNodes().Where(n => HasMatchingSyntaxNode(newDocument, n, oldName));
+                var declaration = declarations.FirstOrDefault();
```

Comment:
Why not loop over these declarations instead of putting everything inside a while loop?